File: Week 7/2. Bank Transaction/BankTransaction.java
import java.io.*;
import java.util.*;

public class BankTransaction {
    static int totalTransactions = 0;
    static int totalMoneyTransactions = 0;
    static Map<String, Integer> fromTransactions = new HashMap<>();
    static Set<String> uniqueAccounts = new TreeSet<>();
    static Map<String, List<String>> graph = new HashMap<>();
    
    public static boolean hasCycleOfLengthK(String start, String current, int k, int depth, Set<String> visited) {
        if (depth == k) 
            return current == start;
        visited.add(current);
        
        if (graph.containsKey(current)) {
            for (String neighbor: graph.get(current)) {
                if (neighbor.equals(start) && depth + 1 == k)
                    return true;
                    
                if (!visited.contains(neighbor)) {
                    if (hasCycleOfLengthK(start, neighbor, k, depth + 1, visited))
                        return true;
                }
            }
        }
        visited.remove(current);
        return false;
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        
        while (true) {
            String line = reader.readLine().trim();
            if (line.equals("#")) break;
            
            String[] parts = line.split(" ");
            String fromAccount = parts[0];
            String toAccount = parts[1];
            int money = Integer.parseInt(parts[2]);
            
            totalTransactions++;
            totalMoneyTransactions += money;
            
            fromTransactions.put(fromAccount, fromTransactions.getOrDefault(fromAccount, 0) + money);
            
            uniqueAccounts.add(fromAccount);
            uniqueAccounts.add(toAccount);
            
            graph.computeIfAbsent(fromAccount, k -> new ArrayList<>()).add(toAccount);
        }
        
        while (true) {
            String line = reader.readLine().trim();
            if (line.equals("#")) break;
            
            String[] parts = line.split(" ");
            String query = parts[0];
            
            switch(query) {
                case ("?number_transactions"):
                    System.out.println(totalTransactions);
                    break;
                
                case ("?total_money_transaction"):
                    System.out.println(totalMoneyTransactions);
                    break;
                
                case ("?list_sorted_accounts"):
                    System.out.println(String.join(" ", uniqueAccounts));
                    break;
                    
                case("?total_money_transaction_from"):
                    String transactionFromAccount = parts[1];
                    System.out.println(fromTransactions.getOrDefault(transactionFromAccount, 0));
                    break;
                    
                case("?inspect_cycle"):
                    String inspectCycleAccount = parts[1];
                    int k = Integer.parseInt(parts[2]);
                    
                    Set<String> visited = new HashSet<>();
                    boolean hasCycle = hasCycleOfLengthK(inspectCycleAccount, inspectCycleAccount, k, 0, visited);
                    System.out.println(hasCycle ? 1 : 0);
            }
        }
    }
}

File: Week 7/2. Bank Transaction/bank_transaction.py
from collections import defaultdict
from typing import List, Set 

total_transactions = 0
total_money_transaction = 0
from_transactions = defaultdict(int)
unique_accounts = set()
graph = defaultdict(list)

def has_cycle_of_length_k(start: str, current: str, k: int, depth: int, visited: Set[str]) -> bool:
    if depth == k:
        return current == start
        
    visited.add(current)
    
    for neighbor in graph[current]:
        if neighbor == start and depth + 1 == k:
            return True
        if neighbor not in visited:
            if has_cycle_of_length_k(start, neighbor, k, depth + 1, visited):
                return True
                
    visited.remove(current)
    return False
    
while True:
    line = input().strip()
    if line == "#":
        break
    
    parts = line.split()
    from_account, to_account, money, time_point, atm = parts[0], parts[1], int(parts[2]), parts[3], parts[4]
    
    total_transactions += 1 
    total_money_transaction += money
    
    from_transactions[from_account] += money
    
    unique_accounts.add(from_account)
    unique_accounts.add(to_account)
    graph[from_account].append(to_account)
    
while True:
    line = input().strip()
    if line == "#":
        break
    
    parts = line.split()
    query = parts[0]
    
    if query == "?number_transactions":
        print(total_transactions)
    elif query == "?total_money_transaction":
        print(total_money_transaction)
    elif query == "?list_sorted_accounts":
        print(" ".join(sorted(unique_accounts)))
    elif query == "?total_money_transaction_from":
        transaction_from_account = parts[1]
        print(from_transactions[transaction_from_account])
    elif query == "?inspect_cycle":
        inspect_cycle_account = parts[1]
        k = int(parts[2])
        
        visited = set()
        has_cycle = has_cycle_of_length_k(inspect_cycle_account, inspect_cycle_account, k, 0, visited)
        print(1 if has_cycle else 0)
